File: python/tree/BinaryTree.py
class Node:
    def __init__(self, data, parent=None):
        self.data = data
        self.parent = parent
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.__root = None
        self.__length = 0

    def __len__(self):
        return self.__length

    def __contains__(self, data):
        def find_data(curr):
            if data == curr.data: return True
            elif data < curr.data:
                if curr.left is None: return False
                else: return find_data(curr.left)
            else:
                if curr.right is None: return False
                else: return find_data(curr.right)

        if self.__root is None: return False
        return find_data(self.__root)

    def __iter__(self):
        def list_gen(curr):
            if curr.left is not None: yield from list_gen(curr.left)
            yield curr.data
            if curr.right is not None: yield from list_gen(curr.right)
        
        if self.__root is None: return iter(())
        return list_gen(self.__root)

    def get_root(self):
        return self.__root
    
    def insert(self, data):
        def do_insertion(curr):
            if data < curr.data:
                if curr.left is None: curr.left = Node(data, curr)
                else: do_insertion(curr.left)
            else:
                if curr.right is None: curr.right = Node(data, curr)
                else: do_insertion(curr.right)

        if self.__root is None: self.__root = Node(data)
        else: do_insertion(self.__root)
        self.__length += 1

    def replace(self, data, new_data):
        self.delete(data)
        self.insert(new_data)
    
    def delete(self, data):
        def replace_with_next_highest(node):
            def get_left(curr):
                while curr.left is not None: curr = curr.left
                return curr

            assert node.right is not None

            if node.right.left is None:
                # print('Leaf node')
                node.data = node.right.data
                if node.right.right is None: node.right = None  # Leaf node
                else:                                           # Only has a right subtree
                    node.right.right.parent = node
                    node.right = node.right.right
            else:
                # print('Tree node')
                next_highest = get_left(node.right)
                node.data = next_highest.data
                if next_highest.right is None: next_highest.parent.left = None  # Leaf node
                else:                                                           # Only has a right subtree
                    next_highest.right.parent = next_highest.parent
                    next_highest.parent.left = next_highest.right
            # print('Replaced {} with {}'.format(data, node.data))
        
        def do_deletion(curr, parent, went_left):
            if data == curr.data:
                # print('Commencing delete of {}'.format(data))
                # If curr is a leaf node
                if curr.left is None and curr.right is None:
                    if went_left: parent.left = None
                    else:         parent.right = None
                # If curr only has a left subtree
                elif curr.right is None:
                    if went_left: parent.left = curr.left
                    else:         parent.right = curr.left
                    curr.left.parent = parent
                # If curr only has a right subtree
                elif curr.left is None:
                    if went_left: parent.left = curr.right
                    else:         parent.right = curr.right
                    curr.right.parent = parent
                # If curr has both subtrees
                else:
                    replace_with_next_highest(curr)
                    
            elif data < curr.data:
                if curr.left is None: raise RuntimeError('Data not in tree')
                else: do_deletion(curr.left, curr, True)
            else:
                if curr.right is None: raise RuntimeError('Data not in tree')
                else: do_deletion(curr.right, curr, False)
        
        if self.__root is None: raise RuntimeError('Deletion on empty tree')
        if data == self.__root.data:
            # print('Deleting root')
            if   self.__root.left == None and self.__root.right == None: self.__root = None
            elif self.__root.right == None:                              self.__root = self.__root.left
            elif self.__root.left == None:                               self.__root = self.__root.right
            else:                                                        replace_with_next_highest(self.__root)
        else: do_deletion(self.__root, None, True)
        self.__length -= 1

File: cpp/tree/binary_tree.cpp
#include <iosfwd>

template <typename T>
struct BinaryTree<T>::TreeNode {
    std::unique_ptr<TreeNode> left{nullptr}, right{nullptr};
    TreeNode* parent{nullptr};
    const T val;

    explicit TreeNode(const T v) : val(v) {}
    explicit TreeNode(const T v, TreeNode *const p) : val(v), parent(p) {}
};

template <typename T>
class BinaryTree<T>::const_iterator {
    const TreeNode* node;
    
public:
    using iterator_category = std::input_iterator_tag;
    using value_type = const T;
    using difference_type = const T;
    using pointer = const T*;
    using reference = const T&;

    explicit const_iterator(TreeNode* n) : node(n) {}

    const_iterator& operator++();
    const_iterator operator++(int);
    bool operator==(const_iterator& other) const { return node == other.node; }
    bool operator!=(const_iterator& other) const { return !(*this == other); }
    reference operator*() const { return node->val; }
};

template <typename T>
typename BinaryTree<T>::const_iterator& BinaryTree<T>::const_iterator::operator++() {
    // Node is the root or has a right child - descend to successor node
    if (!node->parent || node->right) {
        node = node->right.get();
        while (node && node->left) node = node->left.get();

    // Node is on the left - ascend once
    } else if (!node->parent->right || node->parent->right.get() != node) {
        node = node->parent;

    // Node is on the right
    } else {
        // Ascend to the parent of the root of this right branch
        auto currVal{node->val};
        while (node->parent && node->parent->right.get() == node) node = node->parent;
        node = node->parent;
        
        // If we ascend to the root, and it is the current node (node == root) or less than the current node (node == endNode)
        if (node && node->val <= currVal) node = nullptr;
    }
    return *this;
}

template <typename T>
typename BinaryTree<T>::const_iterator BinaryTree<T>::const_iterator::operator++(int) {
    const_iterator retVal{*this};
    ++(*this);
    return retVal;
}

template <typename T>
void BinaryTree<T>::printInOrder(const BinaryTree<T>::TreeNode *const node) const {
    if (!node) return;
    printInOrder(node->left.get());
    std::cout << node->val << ", ";
    printInOrder(node->right.get());
}

template <typename T>
typename BinaryTree<T>::TreeNode *const BinaryTree<T>::find(const T val, BinaryTree<T>::TreeNode *const node) const {
    if (!node)            { return nullptr; }
    if (node->val == val) { return node; }
    if (node->val > val)  { return find(val, node->left.get()); }
    return find(val, node->right.get());
}

template <typename T>
void BinaryTree<T>::insert(const T val, TreeNode *const node) {
    if (node->val > val) {
        if (!node->left) {
            node->left = std::make_unique<TreeNode>(val, node);
            ++_size;
        } else {
            insert(val, node->left.get());
        }
    } else if (node->val < val) {
        if (!node->right) {
            node->right = std::make_unique<TreeNode>(val, node);
            ++_size;
        } else {
            insert(val, node->right.get());
        }
    }
}

template <typename T>
typename BinaryTree<T>::TreeNode *const BinaryTree<T>::erase(const T val, BinaryTree<T>::TreeNode *const node) {
    if (!node) { return nullptr; }
    if (node->val == val) {
        if (_size) --_size;

        // Zero children: return null
        if (!node->left && !node->right) return nullptr;

        // One child: return the non-null child
        if (!node->left)  return node->right.release();
        if (!node->right) return node->left.release();

        // Two children: Hibbard deletion
        // First, find successor node
        TreeNode* prev{node};
        TreeNode* replacement{node->right.get()};
        while (replacement->left) {
            prev = replacement;
            replacement = replacement->left.get();
        }

        // If successor == node.right, only replace left subtree
        if (prev == node) {
            replacement = node->right.release();
            replacement->left = std::unique_ptr<TreeNode>(node->left.release());
            if (replacement->left) replacement->left->parent = replacement;

            return replacement;
        }

        // Else, replace both subtrees and repair prev--x-->repl->right to prev->right
        replacement = prev->left.release();
        prev->left = std::unique_ptr<TreeNode>(replacement->right.release());
        if (prev->left) prev->left->parent = prev;

        replacement->left = std::unique_ptr<TreeNode>(node->left.release());
        replacement->right = std::unique_ptr<TreeNode>(node->right.release());
        if (replacement->left) replacement->left->parent = replacement;
        if (replacement->right) replacement->right->parent = replacement;

        return replacement;
    }
    if (node->val > val) {
        auto retNode{erase(val, node->left.get())};
        if (retNode != node->left.get()) {
            if (retNode) retNode->parent = node;
            node->left = std::unique_ptr<TreeNode>(retNode);
        }
    } else {
        auto retNode{erase(val, node->right.get())};
        if (retNode != node->right.get()) {
            if (retNode) retNode->parent = node;
            node->right = std::unique_ptr<TreeNode>(retNode);
        }
    }
    return node;
}

template <typename T>
void BinaryTree<T>::print() const {
    printInOrder(root.get());
    std::cout << std::endl;
}

template <typename T>
void BinaryTree<T>::insert(const T val) {
    if (!root) {
        root = std::make_unique<TreeNode>(val);
        ++_size;
    } else {
        insert(val, root.get());
    }
}

template <typename T>
void BinaryTree<T>::erase(const T val) {
    TreeNode *const newRoot = erase(val, root.get());
    if (newRoot != root.get()) {
        if (newRoot) newRoot->parent = nullptr;
        root = std::unique_ptr<TreeNode>(newRoot);
    }
};

template <typename T>
void BinaryTree<T>::clear() {
    root.reset();
    _size = 0;
}

template <typename T>
const typename BinaryTree<T>::const_iterator BinaryTree<T>::begin() const {
    if (empty()) return _end;

    auto retNode{root.get()};
    while (retNode->left) retNode = retNode->left.get();
    return const_iterator(retNode);
}
